README.md
# SnowflakeId

[![Java CI](https://github.com/phxql/snowflake-id/actions/workflows/build.yaml/badge.svg)](https://github.com/phxql/snowflake-id/actions/workflows/build.yaml)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/de.mkammerer.snowflake-id/snowflake-id/badge.svg)](https://maven-badges.herokuapp.com/maven-central/de.mkammerer.snowflake-id/snowflake-id)

Generates Twitter-like Snowflake ids.
In short, this is an id scheme to generate unique 64 bit ids which are roughly sortable across multiple systems without
a central
instance. [See this blog post for more details](https://blog.twitter.com/engineering/en_us/a/2010/announcing-snowflake.html).

This project was heavily inspired by [IdGen](https://github.com/RobThree/IdGen) for C#, there's a
great [How it works](https://github.com/RobThree/IdGen#how-it-works) in the readme, too.

The algorithm is implemented in plain Java without any dependencies. All you need is at least Java 11.

Such generated id in binary looks like this (this is 4425020822061056 in decimal):

```
0000000000001111101110001000100001110010001110010000000000000000
||                                            | |
|| Timestamp (16880114830)                    | | Sequence (0)
|                                             |
| Sign bit (always 0)                         | Generator id (1)
```

The structure used for this is 45 bits for the timestamp, 2 for the generator and the remaining 16 for the sequence. This structure can be changed easily, see below for the code. 

The ids really use only 63 bits of the 64 available to circumvent problems with unsigned longs. The generated values are always positive.
Ids from the same generator are monotonically increasing.

## How to use

### Maven

```xml
<dependency>
    <groupId>de.mkammerer.snowflake-id</groupId>
    <artifactId>snowflake-id</artifactId>
    <version>0.0.2</version>
</dependency>
```

### Gradle

```gradle
implementation 'de.mkammerer.snowflake-id:snowflake-id:0.0.2'
```

### Simple example

```java
// generatorId must be unique over all your instances!
int generatorId = 0; 

// use default generator settings
SnowflakeIdGenerator generator = SnowflakeIdGenerator.createDefault(generatorId);

// generate 10 ids
for (int i = 0; i < 10; i++) {
    long id = generator.next();
    System.out.println(id);
}
```

### More configuration options

```java
// Use a custom epoch
TimeSource timeSource = new MonotonicTimeSource(Instant.parse("2020-04-01T00:00:00Z"));
// Use 45 bits for the timestamp, 2 bits for the generator and 16 bits for the sequence
Structure structure = new Structure(45, 2, 16);
// If the sequence number overflows, throw an exception
Options options = new Options(Options.SequenceOverflowStrategy.THROW_EXCEPTION);

// generatorId must be unique over all your instances!
long generatorId = 1;
SnowflakeIdGenerator generator = SnowflakeIdGenerator.createCustom(generatorId, timeSource, structure, options);

// generate 10 ids
for (int i = 0; i < 10; i++) {
    long id = generator.next();
    System.out.println(id);
}
```

### Calculate maximum timestamps, generators, sequence ids and wraparound dates

You can query the `Structure` class to find out the maximum numbers of timestamps, generators, sequence ids and wraparound dates:

```java
TimeSource timeSource = new MonotonicTimeSource(Instant.parse("2020-04-01T00:00:00Z"));

Structure structure = new Structure(45, 2, 16);
System.out.println("Max generators: " + structure.maxGenerators());
System.out.println("Max sequences per ms per generator: " + structure.maxSequenceIds());
System.out.println("Max sequences per ms total: " + structure.maxSequenceIds() * structure.maxGenerators());
System.out.println("Wraparound duration: " + structure.calculateWraparoundDuration(timeSource));
System.out.println("Wraparound date: " + structure.calculateWraparoundDate(timeSource));
```

This prints:

```
Max generators: 4
Max sequences per ms per generator: 65536
Max sequences per ms total: 262144
Wraparound duration: PT9773436H41M28.832S
Wraparound date: 3135-03-14T12:41:28.832Z
```

### Default settings

The default settings for the `Structure` are 41 bits for the timestamp, 10 for the generator id and 12 for the sequence.

This will lead to the following properties:

```
Max generators: 1024
Max sequences per ms per generator: 4096
Max sequences per ms total: 4194304
Wraparound duration: PT610839H47M35.552S
Wraparound date: 2089-09-06T15:47:35.552Z
```

## Building from source

[See here](docs/building.md).

## Changelog?

[See here](CHANGELOG.md).

## License

[LGPLv3](https://www.gnu.org/licenses/lgpl-3.0.html)

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This project adheres
to [Semantic Versioning](http://semver.org/).

## Next version - unreleased

- Improve JavaDoc
- Add new sleep overflow strategy
- Ensure GraalVM compatibility

## [0.0.2] - 2023-01-15

- Add module descriptor. Module is named `de.mkammerer.snowflakeid`
- Switched `synchronized` to `ReentrantLock`
- Fixed bug where `SnowflakeIdGenerator` wasn't thread safe - [#4](https://github.com/phxql/snowflake-id/issues/4)

## [0.0.1] - 2020-10-18

- First official release


src/test/java/de/mkammerer/snowflakeid/test/MockTimeSource.java
package de.mkammerer.snowflakeid.test;

import de.mkammerer.snowflakeid.time.TimeSource;

import java.time.Duration;
import java.time.Instant;

public class MockTimeSource implements TimeSource {
    public static final Instant DEFAULT_EPOCH = Instant.parse("2020-01-01T00:00:00Z");

    private final Instant epoch;
    private long ticks;

    public MockTimeSource(Instant epoch, long ticks) {
        this.epoch = epoch;
        this.ticks = ticks;
    }

    public void setTicks(long ticks) {
        this.ticks = ticks;
    }

    @Override
    public long getTicks() {
        return ticks;
    }

    @Override
    public Duration getTickDuration() {
        return Duration.ofMillis(1);
    }

    @Override
    public Instant getEpoch() {
        return epoch;
    }
}


src/test/java/de/mkammerer/snowflakeid/structure/StructureTest.java
package de.mkammerer.snowflakeid.structure;

import de.mkammerer.snowflakeid.test.MockTimeSource;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
public class StructureTest {
    @Test
    public void timestamp_bits_cant_be_zero() {
        assertThatThrownBy(() ->
            new Structure(0, 31, 31)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("timestampBits");
    }

    @Test
    public void generator_bits_cant_be_zero() {
        assertThatThrownBy(() ->
            new Structure(31, 0, 31)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("generatorBits");
    }

    @Test
    public void generator_bits_cant_more_than_31() {
        assertThatThrownBy(() ->
            new Structure(31, 32, 0)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("generatorBits");
    }

    @Test
    public void sequence_bits_cant_be_zero() {
        assertThatThrownBy(() ->
            new Structure(31, 31, 0)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("sequenceBits");
    }

    @Test
    public void sequence_bits_cant_more_than_31() {
        assertThatThrownBy(() ->
            new Structure(31, 1, 32)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("sequenceBits");
    }

    @Test
    public void parts_must_sum_to_63() {
        assertThatThrownBy(() ->
            new Structure(1, 1, 1)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("must be 63");
    }

    @Test
    public void max_timestamps() {
        Structure structure = new Structure(31, 10, 22);

        assertThat(structure.maxTimestamps()).isEqualTo(2147483648L);
    }

    @Test
    public void max_generators() {
        Structure structure = new Structure(31, 10, 22);

        assertThat(structure.maxGenerators()).isEqualTo(1024L);
    }

    @Test
    public void max_sequence_ids() {
        Structure structure = new Structure(31, 10, 22);

        assertThat(structure.maxSequenceIds()).isEqualTo(4194304L);
    }

    @Test
    public void calculate_wraparound_duration() {
        Structure structure = new Structure(31, 10, 22);

        MockTimeSource mockTimeSource = new MockTimeSource(MockTimeSource.DEFAULT_EPOCH, 0);

        assertThat(structure.calculateWraparoundDuration(mockTimeSource)).isEqualTo(Duration.ofMillis(2147483648L));
    }

    @Test
    public void calculate_wraparound_date() {
        Structure structure = new Structure(31, 10, 22);

        MockTimeSource mockTimeSource = new MockTimeSource(Instant.parse("2020-01-01T00:00:00Z"), 0);

        // epoch + 2147483648 milliseconds
        assertThat(structure.calculateWraparoundDate(mockTimeSource)).isEqualTo(Instant.parse("2020-01-25T20:31:23.648Z"));
    }
}

src/test/java/de/mkammerer/snowflakeid/SnowflakeIdGeneratorTest.java
package de.mkammerer.snowflakeid;

import de.mkammerer.snowflakeid.options.Options;
import de.mkammerer.snowflakeid.structure.Structure;
import de.mkammerer.snowflakeid.test.MockTimeSource;
import de.mkammerer.snowflakeid.time.MonotonicTimeSource;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import static org.assertj.core.api.Assertions.*;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
public class SnowflakeIdGeneratorTest {
    @ParameterizedTest
    @EnumSource(value = Options.SequenceOverflowStrategy.class, names = "THROW_EXCEPTION", mode = EnumSource.Mode.EXCLUDE)
    public void generate_unique(Options.SequenceOverflowStrategy strategy) {
        // Use two generators to generate 20000 ids and check them for uniqueness and increasing order
        SnowflakeIdGenerator generator0 = SnowflakeIdGenerator.createCustom(0, MonotonicTimeSource.createDefault(), Structure.createDefault(), new Options(strategy));
        SnowflakeIdGenerator generator1 = SnowflakeIdGenerator.createCustom(1, MonotonicTimeSource.createDefault(), Structure.createDefault(), new Options(strategy));
        int count = 10_000;

        Set<Long> ids = new HashSet<>(count);

        long lastGen0Id = -1;
        long lastGen1Id = -1;

        for (int i = 0; i < count; i++) {
            long gen0id = generator0.next();
            long gen1id = generator1.next();

            // All ids from the same generator are increasing
            assertThat(gen0id).isGreaterThan(lastGen0Id);
            lastGen0Id = gen0id;
            assertThat(gen1id).isGreaterThan(lastGen1Id);
            lastGen1Id = gen1id;

            if (!ids.add(gen0id)) {
                fail(gen0id + " is a duplicate");
            }
            if (!ids.add(gen1id)) {
                fail(gen1id + " is a duplicate");
            }
        }

        assertThat(ids).hasSize(count * 2);
    }

    @Test
    public void forbid_negative_generator_id() {
        assertThatThrownBy(() ->
            SnowflakeIdGenerator.createDefault(-1)
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("generatorId");
    }

    @Test
    public void forbid_too_big_generator_id() {
        assertThatThrownBy(() ->
            // Maximum generator id is 7
            SnowflakeIdGenerator.createCustom(1024, MonotonicTimeSource.createDefault(), new Structure(50, 3, 10), Options.createDefault())
        ).isInstanceOf(IllegalArgumentException.class).hasMessageContaining("generatorId");
    }

    @Test
    public void sequence_overflow_exception() {
        // We use 1 bit for the sequence, this should overflow pretty fast!
        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createCustom(0, MonotonicTimeSource.createDefault(), new Structure(50, 12, 1), new Options(Options.SequenceOverflowStrategy.THROW_EXCEPTION));

        assertThatThrownBy(() -> {
            for (int i = 0; i < 10000; i++) {
                sut.next();
            }
        }).isInstanceOf(IllegalStateException.class).hasMessageContaining("Sequence overflow");
    }

    @Test
    public void sequence_overflow_spin_wait() {
        // We use 1 bit for the sequence, this should overflow pretty fast!
        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createCustom(0, MonotonicTimeSource.createDefault(), new Structure(50, 12, 1), new Options(Options.SequenceOverflowStrategy.SPIN_WAIT));

        assertThatCode(() -> {
            for (int i = 0; i < 10; i++) {
                sut.next();
            }
        }).doesNotThrowAnyException();
    }

    @Test
    public void sequence_overflow_sleep() {
        // We use 1 bit for the sequence, this should overflow pretty fast!
        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createCustom(0, MonotonicTimeSource.createDefault(), new Structure(50, 12, 1), new Options(Options.SequenceOverflowStrategy.SLEEP));

        assertThatCode(() -> {
            for (int i = 0; i < 10; i++) {
                sut.next();
            }
        }).doesNotThrowAnyException();
    }

    @Test
    public void protect_against_negative_ticks() {
        MockTimeSource mockTimeSource = new MockTimeSource(MockTimeSource.DEFAULT_EPOCH, -1);

        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createCustom(0, mockTimeSource, Structure.createDefault(), Options.createDefault());

        assertThatThrownBy(sut::next).isInstanceOf(IllegalStateException.class).hasMessageContaining("negative ticks");
    }

    @Test
    public void protect_against_clock_moved_backwards() {
        MockTimeSource mockTimeSource = new MockTimeSource(MockTimeSource.DEFAULT_EPOCH, 1);

        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createCustom(0, mockTimeSource, Structure.createDefault(), Options.createDefault());

        mockTimeSource.setTicks(2);
        sut.next();

        mockTimeSource.setTicks(1);
        assertThatThrownBy(sut::next).isInstanceOf(IllegalStateException.class).hasMessageContaining("moved backwards");

    }

    @Test
    public void is_thread_safe() throws ExecutionException, InterruptedException {
        int workers = 100;
        int ids = 100_000;

        SnowflakeIdGenerator sut = SnowflakeIdGenerator.createDefault(1);
        ExecutorService executorService = Executors.newFixedThreadPool(workers);
        try {
            List<Future<Long>> futures = new ArrayList<>();
            for (int i = 0; i < ids; i++) {
                futures.add(executorService.submit(sut::next));
            }

            Set<Long> generatedIds = new HashSet<>(ids);
            for (Future<Long> future : futures) {
                Long id = future.get();
                if (!generatedIds.add(id)) {
                    fail(id + " is a duplicate");
                }
            }
            assertThat(generatedIds).hasSize(ids);
        } finally {
            executorService.shutdownNow();
        }
    }
}

src/main/java/de/mkammerer/snowflakeid/options/Options.java
package de.mkammerer.snowflakeid.options;

import java.util.Objects;

/**
 * Options.
 *
 * @author Moritz Halbritter
 */
public class Options {
    private final SequenceOverflowStrategy sequenceOverflowStrategy;

    /**
     * Constructor.
     *
     * @param sequenceOverflowStrategy the sequence overflow strategy
     */
    public Options(SequenceOverflowStrategy sequenceOverflowStrategy) {
        this.sequenceOverflowStrategy = Objects.requireNonNull(sequenceOverflowStrategy, "sequenceOverflowStrategy");
    }

    /**
     * Returns the sequence overflow strategy.
     *
     * @return the sequence overflow strategy
     */
    public SequenceOverflowStrategy getSequenceOverflowStrategy() {
        return sequenceOverflowStrategy;
    }

    /**
     * Creates options with default settings.
     * <p>
     * If a sequence overflow occurs, uses spin wait to wait for the next timestamp.
     *
     * @return options
     */
    public static Options createDefault() {
        return new Options(SequenceOverflowStrategy.SPIN_WAIT);
    }

    @Override
    public String toString() {
        return "Options{" +
            "sequenceOverflowStrategy=" + sequenceOverflowStrategy +
            '}';
    }

    /**
     * Sequence overflow strategy.
     */
    public enum SequenceOverflowStrategy {
        /**
         * Throws an exception if a sequence overflow occurs.
         */
        THROW_EXCEPTION,
        /**
         * Spin waits for the next sequence if a sequence overflow occurs.
         */
        SPIN_WAIT,
        /**
         * Sleeps until the next sequence if a sequence overflow occurs.
         */
        SLEEP
    }
}


src/main/java/de/mkammerer/snowflakeid/structure/Structure.java
package de.mkammerer.snowflakeid.structure;

import de.mkammerer.snowflakeid.time.TimeSource;

import java.time.Duration;
import java.time.Instant;

/**
 * Id structure.
 *
 * @author Moritz Halbritter
 */
public class Structure {
    private final int timestampBits;
    private final int generatorBits;
    private final int sequenceBits;

    /**
     * Constructor.
     *
     * @param timestampBits the bits used for the timestamp. Must be greater than 0
     * @param generatorBits the bits used for the generator. Must be between 1 (inclusive) and 31 (inclusive)
     * @param sequenceBits  the bits used for the sequence. Must be between 1 (inclusive) and 31 (inclusive)
     * @throws IllegalArgumentException if an argument is invalid, or if the bits don't add up to 63
     */
    public Structure(int timestampBits, int generatorBits, int sequenceBits) {
        if (timestampBits < 1) {
            throw new IllegalArgumentException("timestampBits must no be <= 0, but was " + timestampBits);
        }
        if (generatorBits < 1 || generatorBits > 31) {
            throw new IllegalArgumentException("generatorBits must be between 1 (inclusive) and 31 (inclusive), but was " + generatorBits);
        }
        if (sequenceBits < 1 || sequenceBits > 31) {
            throw new IllegalArgumentException("sequenceBits must be between 1 (inclusive) and 31 (inclusive), but was " + sequenceBits);
        }

        int sum = timestampBits + generatorBits + sequenceBits;
        if (sum != 63) {
            throw new IllegalArgumentException("timestampBits + generatorBits + sequenceBits must be 63, but was " + sum);
        }

        this.timestampBits = timestampBits;
        this.generatorBits = generatorBits;
        this.sequenceBits = sequenceBits;
    }

    /**
     * Returns the bits used for the timestamp.
     *
     * @return the bits used for the timestamp
     */
    public int getTimestampBits() {
        return timestampBits;
    }

    /**
     * Returns the bits used for the generator.
     *
     * @return the bits used for the generator
     */
    public int getGeneratorBits() {
        return generatorBits;
    }

    /**
     * Returns the bits used for the sequence.
     *
     * @return the bits used for the sequence
     */
    public int getSequenceBits() {
        return sequenceBits;
    }

    /**
     * Returns the maximum number of generators.
     *
     * @return the maximum number of generators
     */
    public long maxGenerators() {
        return 1L << generatorBits;
    }

    /**
     * Returns the maximum number of sequence ids.
     *
     * @return the maximum number of sequence ids
     */
    public long maxSequenceIds() {
        return 1L << sequenceBits;
    }

    /**
     * Returns the maximum number of timestamps.
     *
     * @return the maximum number of timestamps
     */
    public long maxTimestamps() {
        return 1L << timestampBits;
    }

    /**
     * Calculates when the sequence ids will wrap around
     *
     * @param timeSource the used time source
     * @return the wrap around duration
     */
    public Duration calculateWraparoundDuration(TimeSource timeSource) {
        return timeSource.getTickDuration().multipliedBy(maxTimestamps());
    }

    /**
     * Calculates when the sequence ids will wrap around
     *
     * @param timeSource the used time source
     * @return the wrap around instant
     */
    public Instant calculateWraparoundDate(TimeSource timeSource) {
        return timeSource.getEpoch().plus(calculateWraparoundDuration(timeSource));
    }

    /**
     * Creates a structure with default settings.
     * <p>
     * Uses 41 bits for the timestamp, 10 for the generator id and 12 for the sequence.
     *
     * @return structure
     */
    public static Structure createDefault() {
        return new Structure(41, 10, 12);
    }

    @Override
    public String toString() {
        return "Structure{" +
            "timestampBits=" + timestampBits +
            ", generatorBits=" + generatorBits +
            ", sequenceBits=" + sequenceBits +
            '}';
    }
}


src/main/java/de/mkammerer/snowflakeid/SnowflakeIdGenerator.java
package de.mkammerer.snowflakeid;

import de.mkammerer.snowflakeid.options.Options;
import de.mkammerer.snowflakeid.structure.Structure;
import de.mkammerer.snowflakeid.time.MonotonicTimeSource;
import de.mkammerer.snowflakeid.time.TimeSource;

import java.util.Objects;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Generates snowflake ids. This class is thread safe.
 *
 * @author Moritz Halbritter
 */
public class SnowflakeIdGenerator {
    /**
     * Lock for sequence and lastTimestamp.
     */
    private final Lock lock = new ReentrantLock();

    // Stuff which is set in the constructor
    private final long generatorId;
    private final TimeSource timeSource;
    private final Structure structure;
    private final Options options;

    // precalculated variables for bit magic
    private final long maxSequence;
    private final long maskTime;
    private final int shiftTime;
    private final int shiftGenerator;

    /**
     * Tracks the last generated timestamp.
     */
    private long lastTimestamp = -1;
    /**
     * Sequence number, unique per timestamp.
     */
    private long sequence = 0;

    // Structure:
    // time || generator || sequence
    private SnowflakeIdGenerator(long generatorId, TimeSource timeSource, Structure structure, Options options) {
        this.timeSource = Objects.requireNonNull(timeSource, "timeSource");
        this.structure = Objects.requireNonNull(structure, "structure");
        this.options = Objects.requireNonNull(options, "options");

        if (generatorId < 0 || generatorId >= structure.maxGenerators()) {
            throw new IllegalArgumentException("generatorId must be between 0 (inclusive) and " + structure.maxGenerators() + " (exclusive), but was " + generatorId);
        }

        this.generatorId = generatorId;

        maskTime = calculateMask(structure.getTimestampBits());
        maxSequence = calculateMask(structure.getSequenceBits());
        shiftTime = structure.getGeneratorBits() + structure.getSequenceBits();
        shiftGenerator = structure.getSequenceBits();
    }

    /**
     * Generates the next id.
     *
     * @return next id
     * @throws IllegalStateException if some invariant has been broken, e.g. the clock moved backwards or a sequence overflow occurred
     */
    public long next() {
        lock.lock();
        try {
            long ticks = timeSource.getTicks();
            if (ticks < 0) {
                throw new IllegalStateException("Clock gave negative ticks");
            }
            long timestamp = ticks & maskTime;

            // Guard against non-monotonic clocks
            if (timestamp < lastTimestamp) {
                throw new IllegalStateException("Timestamp moved backwards or wrapped around");
            }

            if (timestamp == lastTimestamp) {
                // Same timeslot
                if (sequence >= maxSequence) {
                    handleSequenceOverflow();
                    return next();
                }
                sequence++;
            } else {
                // other timeslot, reset sequence
                sequence = 0;
                lastTimestamp = timestamp;
            }

            return (timestamp << shiftTime) + (generatorId << shiftGenerator) + sequence;
        } finally {
            lock.unlock();
        }
    }

    /**
     * Returns the generator id.
     *
     * @return the generator id
     */
    public long getGeneratorId() {
        return generatorId;
    }

    /**
     * Returns the time source.
     *
     * @return the time source
     */
    public TimeSource getTimeSource() {
        return timeSource;
    }

    /**
     * Returns the options.
     *
     * @return the options
     */
    public Options getOptions() {
        return options;
    }

    /**
     * Returns the structure.
     *
     * @return the structure
     */
    public Structure getStructure() {
        return structure;
    }

    /**
     * Creates a custom snowflake id generator.
     *
     * @param generatorId the id of the generator. Must be unique across all instances
     * @param timeSource  the timesource to use
     * @param structure   the id structure
     * @param options     the options
     * @return the created snowflake id generator
     */
    public static SnowflakeIdGenerator createCustom(long generatorId, TimeSource timeSource, Structure structure, Options options) {
        return new SnowflakeIdGenerator(generatorId, timeSource, structure, options);
    }

    /**
     * Creates a generator with default settings.
     * <p>
     * Uses 2020-01-01T00:00:00Z as epoch, 41 bits for the timestamp, 10 for the generator id and 12 for the sequence. If a
     * sequence overflow occurs, uses spin wait to wait for the next timestamp.
     *
     * @param generatorId the id of the generator. Must be unique across all instances
     * @return generator
     */
    public static SnowflakeIdGenerator createDefault(int generatorId) {
        return new SnowflakeIdGenerator(generatorId, MonotonicTimeSource.createDefault(), Structure.createDefault(), Options.createDefault());
    }

    private void handleSequenceOverflow() {
        switch (this.options.getSequenceOverflowStrategy()) {
            case THROW_EXCEPTION:
                throw new IllegalStateException("Sequence overflow");
            case SPIN_WAIT:
                spinWaitForNextTick(lastTimestamp);
                break;
            case SLEEP:
                sleepForTickDuration();
                break;
            default:
                throw new AssertionError("Unexpected enum value: " + this.options.getSequenceOverflowStrategy());
        }
    }

    private void sleepForTickDuration() {
        try {
            Thread.sleep(timeSource.getTickDuration().toMillis());
        } catch (InterruptedException e) {
            // Do nothing
        }
    }

    private void spinWaitForNextTick(long lastTimestamp) {
        long timestamp;
        do {
            Thread.onSpinWait();
            timestamp = timeSource.getTicks() & maskTime;
        } while (timestamp == lastTimestamp);
    }

    private long calculateMask(int bits) {
        return (1L << bits) - 1;
    }

    @Override
    public String toString() {
        return "SnowflakeIdGenerator{" +
            "generatorId=" + generatorId +
            ", timeSource=" + timeSource +
            ", structure=" + structure +
            ", options=" + options +
            '}';
    }
}


src/main/java/de/mkammerer/snowflakeid/time/MonotonicTimeSource.java
package de.mkammerer.snowflakeid.time;

import java.time.Duration;
import java.time.Instant;
import java.util.Objects;

/**
 * A monotonic time source, based on {@link System#nanoTime()}.
 *
 * @author Moritz Halbritter
 */
public class MonotonicTimeSource implements TimeSource {
    private final long start;
    private final long offset;
    private final Instant epoch;

    /**
     * Constructor.
     *
     * @param epoch the epoch
     */
    public MonotonicTimeSource(Instant epoch) {
        this.epoch = Objects.requireNonNull(epoch, "epoch");

        // Offset elapsed time by this amount (creation time of this time source since epoch)
        offset = Instant.now().toEpochMilli() - epoch.toEpochMilli();
        // Record creation of this time source in milliseconds
        start = System.nanoTime() / 1_000_000;
    }

    @Override
    public long getTicks() {
        return offset + elapsed();
    }

    @Override
    public Duration getTickDuration() {
        return Duration.ofMillis(1);
    }

    @Override
    public Instant getEpoch() {
        return epoch;
    }

    /**
     * Creates a time source with default settings.
     * <p>
     * Uses 2020-01-01T00:00:00Z as epoch.
     *
     * @return time source
     */
    public static MonotonicTimeSource createDefault() {
        // 2020-01-01T00:00:00Z
        return new MonotonicTimeSource(Instant.ofEpochMilli(1577836800000L));
    }

    private long elapsed() {
        // Calculate elapsed time since creation of this time source in milliseconds
        return (System.nanoTime() / 1_000_000) - start;
    }

    @Override
    public String toString() {
        return "MonotonicTimeSource{" +
            "epoch=" + epoch +
            '}';
    }
}


src/main/java/de/mkammerer/snowflakeid/Main.java
package de.mkammerer.snowflakeid;

import de.mkammerer.snowflakeid.options.Options;
import de.mkammerer.snowflakeid.structure.Structure;
import de.mkammerer.snowflakeid.time.MonotonicTimeSource;
import de.mkammerer.snowflakeid.time.TimeSource;

import java.time.Instant;
import java.util.HashSet;
import java.util.Set;

public class Main {
    // TODO: Remove, maybe move to a test
    public static void main(String[] args) {
        TimeSource timeSource = new MonotonicTimeSource(Instant.parse("2020-04-01T00:00:00Z"));

        Structure structure = new Structure(45, 2, 16);
        System.out.println("Max generators: " + structure.maxGenerators());
        System.out.println("Max sequences per ms per generator: " + structure.maxSequenceIds());
        System.out.println("Max sequences per ms total: " + structure.maxSequenceIds() * structure.maxGenerators());
        System.out.println("Wraparound duration: " + structure.calculateWraparoundDuration(timeSource));
        System.out.println("Wraparound date: " + structure.calculateWraparoundDate(timeSource));

        Options options = Options.createDefault();

        SnowflakeIdGenerator gen0 = SnowflakeIdGenerator.createCustom(0, timeSource, structure, options);
        SnowflakeIdGenerator gen1 = SnowflakeIdGenerator.createCustom(1, timeSource, structure, options);
        SnowflakeIdGenerator gen2 = SnowflakeIdGenerator.createCustom(2, timeSource, structure, options);
        SnowflakeIdGenerator gen3 = SnowflakeIdGenerator.createCustom(3, timeSource, structure, options);

        Set<Long> set = new HashSet<>();

        for (int i = 0; i < 100_000; i++) {
            addToSet(set, gen0.next());
            addToSet(set, gen1.next());
            addToSet(set, gen2.next());
            addToSet(set, gen3.next());
        }

        // Print first id
        System.out.println(set.iterator().next());
    }

    private static void addToSet(Set<Long> set, long id) {
        // Check for uniqueness
        if (!set.add(id)) {
            throw new IllegalStateException(id + " is a duplicate");
        }
    }
}
